### PR TITLE
[Frontend/Runtime] Add support for nested Hamiltonian observables

### PR DESCRIPTION
**Pull request by maliasadi**
Update the runtime to support measurement processes with nested Hamiltonian observables. In runtime, we used to check the type of given observables for constructing the `HamiltonianObs` data-type and obs-id in the Observable Manager. This is now passed to the device to check whether those observables are supported in the `Hamiltonian` method or not.  

For Lightning simulators, these devices support nested Hamiltonians and will raise errors for unsupported observables while we still check the validity of given observables for OpenQasmDevice in the OpenQasmObsManager class as we have a limited support for this particular device at runtime

You can now JIT-compile the following circuit:  

  ``` python
  @qjit
  @qml.qnode(qml.device("lightning.qubit", wires=3))
  def circuit(x, y, coeffs1, coeffs2):
      qml.RX(x, wires=0)
      qml.RX(y, wires=1)
      qml.RY(x + y, wires=2)

      obs = [
          qml.PauliX(0) @ qml.PauliZ(1),
          qml.Hamiltonian(coeffs1, [qml.PauliZ(0) @ qml.Hadamard(2)]),
      ]

      return qml.var(qml.Hamiltonian(coeffs2, obs))
  ```

[sc-41351]

### CODE REVIEW

File: doc/changelog.md [hunk 1]
```diff
@@ -143,6 +143,24 @@
...
+      qml.RX(y, wires=1)
+      qml.RY(x + y, wires=2)
+
+      obs = [
+          qml.PauliX(0) @ qml.PauliZ(1),
+          qml.Hamiltonian(coeffs1, [qml.PauliZ(0) @ qml.Hadamard(2)]),
+      ]
+
+      return qml.var(qml.Hamiltonian(coeffs2, obs))
+  ```
```
Comment (josh146) [inline doc/changelog.md:163]: Quick question @maliasadi -- does this feature also work for the new op arithmetic you are working on in the other open PR?

Comment (maliasadi) [reply doc/changelog.md:163]: Yes @josh146. This works for the new arithmetic operators and it's basically the last piece of the puzzle. After merging this PR, we can merge PR #250.

Comment (maliasadi) [reply doc/changelog.md:163]: Yes @josh146. This works for the new operators too and it's technically the last piece of the puzzle. After merging this PR, we should be able to merge #250.

**Comment by maliasadi** [comment]
> What happens if they are not supported by the device? What kind of error message is given?

@erick-xanadu for Lightning simulators, these devices will raise errors for unsupported observables while we still check the validity of given observables for `OpenQasmDevice` in the `OpenQasmObsManager` class as we have a limited support for this particular device at runtime.

**Comment by maliasadi** [comment]
> So was this support already present in Lightning? Is that how they support the new pennylane arithmatic?

Yes, it's supported in Lightning sims as mentioned and that's the reason they don't need to deal with the generated nested Hamiltonians with either the current or new arithmetic ops.

File: doc/changelog.md [hunk 2]
```diff
@@ -143,6 +143,24 @@
       return qml.expval(qml.Hamiltonian(coeffs, obs))
   ```
 
+* Add support for nested Hamiltonian observables.
```
Comment (dime10) [inline doc/changelog.md:146]: I think this PR (and 248 just above) would fit better in the *Improvements* section than in `New features* section. I know it's not always clear which is which but here the impact is fairly small compared to something like the Enzyme support.

**Review by dime10 — APPROVED** [review]
Thanks @maliasadi 👍

**Pull request merged**